File: python/src/threads/_json_schema.py
"""The JSON Schema evaluator for the keywords threads checks at runtime: the cross-field rules the
model generator leaves to `StrictModel`, and pinned output schemas (semantic rule 20), which are
what a Pydantic or Zod schema writes: constraints, formats, enums, nested and recursive
models by `$ref`. spec/schema/README.md defines each keyword; validate/json-schema.ts is the
TypeScript reader of the same definition.

`holds` raises TypeError on a keyword it can't check rather than skipping it: a rule that can't be
checked must never pass silently. `unchecked` names such a keyword anywhere in a schema, so an
output model that uses one is refused at setup instead of failing a run.
"""

import operator
import re
from collections.abc import Callable, Mapping, Sequence, Sized
from typing import Final, TypeGuard

from pydantic import JsonValue

from threads._json_formats import FORMATS, multiple_of

type Root = Mapping[str, JsonValue]
"""The whole schema document: `$ref`s resolve against it."""
type _Check = Callable[[JsonValue, object, Root], bool]

_ANNOTATIONS: Final = frozenset(
    {"description", "title", "default", "examples", "$defs", "$schema", "$comment"}
)
"""Keywords that describe a value and never constrain it."""
_DEFS: Final = "#/$defs/"


def holds(schema: JsonValue, value: object) -> bool:
    """Whether `value` satisfies `schema`. A `$ref` is `#` (the whole schema) or names one of
    its `$defs`."""
    return _holds(schema, value, schema if isinstance(schema, dict) else {})


def _holds(schema: JsonValue, value: object, root: Root) -> bool:
    if isinstance(schema, bool):
        return schema
    if not isinstance(schema, dict):
        raise TypeError(f"not a schema: {schema!r}")
    if "if" in schema:
        branch = schema.get("then") if _holds(schema["if"], value, root) else schema.get("else")
        if branch is not None and not _holds(branch, value, root):
            return False
    if "additionalProperties" in schema and not _additional(schema, value, root):
        return False
    for keyword, argument in schema.items():
        if keyword in ("if", "then", "else", "additionalProperties") or keyword in _ANNOTATIONS:
            continue
        check = _CHECKS.get(keyword)
        if check is None:
            raise TypeError(f"unsupported schema keyword {keyword!r}")
        if not check(argument, value, root):
            return False
    return True


def conforms(schema: Mapping[str, JsonValue], value: JsonValue) -> bool:
    """Semantic rule 20: whether an output value satisfies the pinned output schema. A keyword
    this reader can't check fails closed rather than accepting unchecked output."""
    try:
        return holds(dict(schema), value)
    except TypeError:
        return False


def unchecked(schema: JsonValue) -> str | None:
    """The first keyword `holds` can't check anywhere in `schema` (or a `$ref` it can't
    follow, or a pattern it can't compile), else None."""
    try:
        _walk(schema)
    except TypeError as error:
        return str(error)
    return None


_SCHEMA_ARG: Final = frozenset({"not", "if", "then", "else", "items", "additionalProperties"})
_LIST_ARG: Final = frozenset({"allOf", "anyOf", "oneOf"})
_MAP_ARG: Final = frozenset({"properties", "$defs"})


def _walk(schema: JsonValue) -> None:
    if isinstance(schema, bool):
        return
    if not isinstance(schema, dict):
        raise TypeError(f"not a schema: {schema!r}")
    for keyword, argument in schema.items():
        _admit(keyword, argument)
        for sub in _subschemas(keyword, argument):
            _walk(sub)


def _admit(keyword: str, argument: JsonValue) -> None:
    """Raises TypeError for a keyword `holds` can't check."""
    if keyword == "pattern":
        _pattern(argument)
    elif keyword == "$ref":
        _name(argument)
    elif keyword == "format" and argument not in FORMATS:
        raise TypeError(f"unsupported format {argument!r}")
    elif keyword not in _CHECKS and keyword not in _ANNOTATIONS and keyword not in _SCHEMA_ARG:
        raise TypeError(f"unsupported schema keyword {keyword!r}")


def _subschemas(keyword: str, argument: JsonValue) -> Sequence[JsonValue]:
    if keyword in _SCHEMA_ARG:
        return (argument,)
    if keyword in _LIST_ARG:
        return _schemas(argument)
    if keyword in _MAP_ARG:
        return tuple(_map(argument).values())
    return ()


def is_object(value: object) -> TypeGuard[Mapping[str, object]]:
    return isinstance(value, dict)


def _is_array(value: object) -> TypeGuard[Sequence[object]]:
    return isinstance(value, list)


def _is_number(value: object) -> TypeGuard[int | float]:
    return isinstance(value, int | float) and not isinstance(value, bool)


def _schemas(argument: JsonValue) -> Sequence[JsonValue]:
    if not isinstance(argument, list):
        raise TypeError(f"expected a list of schemas: {argument!r}")
    return argument


def _map(argument: JsonValue) -> Mapping[str, JsonValue]:
    if not isinstance(argument, dict):
        raise TypeError(f"expected a map of schemas: {argument!r}")
    return argument


def json_equal(expected: JsonValue, value: object) -> bool:
    # JSON true is not 1 and 1.0 is not the integer 1, though Python's
    # True == 1 == 1.0.
    return type(expected) is type(value) and expected == value


def _name(ref: JsonValue) -> str | None:
    """The `$defs` entry a local reference names, or None for `#`, the whole schema; any other
    reference can't be followed."""
    if ref == "#":
        return None
    if not isinstance(ref, str) or not ref.startswith(_DEFS):
        raise TypeError(f"unsupported $ref {ref!r}")
    return ref.removeprefix(_DEFS)


def _ref(argument: JsonValue, value: object, root: Root) -> bool:
    name = _name(argument)
    if name is None:
        return _holds(dict(root), value, root)
    defs = _map(root.get("$defs", {}))
    if name not in defs:
        raise TypeError(f"$ref to a missing definition {name!r}")
    return _holds(defs[name], value, root)


def _format(argument: JsonValue, value: object, _root: Root) -> bool:
    check = FORMATS.get(argument) if isinstance(argument, str) else None
    if check is None:
        raise TypeError(f"unsupported format {argument!r}")
    return not isinstance(value, str) or check(value)


def _multiple(argument: JsonValue, value: object, _root: Root) -> bool:
    if not _is_number(argument):
        raise TypeError(f"multipleOf is a number: {argument!r}")
    return not _is_number(value) or multiple_of(value, argument)


def _pattern(argument: JsonValue) -> re.Pattern[str]:
    # ECMA-262 without the u flag: \d, \w and \s are ASCII.
    if not isinstance(argument, str):
        raise TypeError(f"a pattern is a string: {argument!r}")
    try:
        return re.compile(argument, re.ASCII)
    except re.error as error:
        raise TypeError(f"pattern {argument!r} doesn't compile: {error}") from None


def _matches(argument: JsonValue, value: object, _root: Root) -> bool:
    return not isinstance(value, str) or _pattern(argument).search(value) is not None


def _required(argument: JsonValue, value: object, _root: Root) -> bool:
    return not is_object(value) or all(key in value for key in _schemas(argument))


def _properties(argument: JsonValue, value: object, root: Root) -> bool:
    return not is_object(value) or all(
        _holds(sub, value[key], root) for key, sub in _map(argument).items() if key in value
    )


def _additional(schema: dict[str, JsonValue], value: object, root: Root) -> bool:
    if not is_object(value):
        return True
    declared = _map(schema.get("properties", {}))
    extra = schema["additionalProperties"]
    return all(_holds(extra, value[key], root) for key in value if key not in declared)


def _items(argument: JsonValue, value: object, root: Root) -> bool:
    return not _is_array(value) or all(_holds(argument, item, root) for item in value)


def _is_integer(value: object) -> bool:
    if isinstance(value, float):
        return value.is_integer()
    return isinstance(value, int) and not isinstance(value, bool)


_JSON_TYPES: Mapping[str, Callable[[object], bool]] = {
    "object": is_object,
    "array": _is_array,
    "string": lambda value: isinstance(value, str),
    "boolean": lambda value: isinstance(value, bool),
    "null": lambda value: value is None,
    "integer": _is_integer,
    "number": _is_number,
}


def _type(argument: JsonValue, value: object, _root: Root) -> bool:
    names = argument if isinstance(argument, list) else [argument]
    for name in names:
        if not isinstance(name, str) or name not in _JSON_TYPES:
            raise TypeError(f"unknown JSON type {name!r}")
    return any(_JSON_TYPES[str(name)](value) for name in names)


type _Compare = Callable[[float, float], bool]


def _bound(compare: _Compare) -> _Check:
    """A bound on a number; any other value passes."""

    def check(argument: JsonValue, value: object, _root: Root) -> bool:
        if not _is_number(argument):
            raise TypeError(f"a bound is a number: {argument!r}")
        return not _is_number(value) or compare(value, argument)

    return check


def _size(of: type[Sized], compare: _Compare) -> _Check:
    """A bound on the length of a string (in code points, as JSON Schema counts), an array or
    an object; any other value passes."""

    def check(argument: JsonValue, value: object, _root: Root) -> bool:
        if not _is_number(argument):
            raise TypeError(f"a length bound is a number: {argument!r}")
        return not isinstance(value, of) or compare(len(value), argument)

    return check


_CHECKS: Mapping[str, _Check] = {
    "not": lambda argument, value, root: not _holds(argument, value, root),
    "allOf": lambda argument, value, root: all(_holds(s, value, root) for s in _schemas(argument)),
    "anyOf": lambda argument, value, root: any(_holds(s, value, root) for s in _schemas(argument)),
    "oneOf": lambda argument, value, root: (
        sum(_holds(s, value, root) for s in _schemas(argument)) == 1
    ),
    "const": lambda argument, value, _root: json_equal(argument, value),
    "enum": lambda argument, value, _root: any(json_equal(e, value) for e in _schemas(argument)),
    "required": _required,
    "properties": _properties,
    "type": _type,
    "items": _items,
    "$ref": _ref,
    "pattern": _matches,
    "format": _format,
    "multipleOf": _multiple,
    "minimum": _bound(operator.ge),
    "maximum": _bound(operator.le),
    "exclusiveMinimum": _bound(operator.gt),
    "exclusiveMaximum": _bound(operator.lt),
    "minLength": _size(str, operator.ge),
    "maxLength": _size(str, operator.le),
    "minItems": _size(list, operator.ge),
    "maxItems": _size(list, operator.le),
    "minProperties": _size(dict, operator.ge),
}

File: typescript/packages/core/src/validate/json-schema.ts
import { FORMATS, multipleOf } from "./formats";

// The JSON Schema evaluator semantic rule 20 checks an output value with: the keyword set that
// spec/schema/README.md ("Output schemas") defines, the same as the Python reader
// (threads/_json_schema.py). Never a library's reading of JSON Schema, so both languages agree.
// A keyword it can't check throws TypeError rather than passing; `unchecked` names one anywhere
// in a schema, so an output schema that uses it is refused at setup instead of failing a run.

type Json = Readonly<Record<string, unknown>>;
type Check = (argument: unknown, value: unknown, root: Json) => boolean;

const ANNOTATIONS: ReadonlySet<string> = new Set([
  "description",
  "title",
  "default",
  "examples",
  "$defs",
  "$schema",
  "$comment",
]);
const DEFS = "#/$defs/";

/** Whether `value` satisfies `schema`. A `$ref` is `#` or names one of its `$defs`. */
export function holds(schema: unknown, value: unknown): boolean {
  return holdsIn(schema, value, isObject(schema) ? schema : {});
}

/**
 * Semantic rule 20: whether an output value satisfies the pinned output schema. A keyword this
 * reader can't check fails closed rather than accepting unchecked output.
 */
export function conforms(schema: unknown, value: unknown): boolean {
  try {
    return holds(schema, value);
  } catch (error) {
    if (error instanceof TypeError) return false;
    throw error;
  }
}

/** The first keyword `holds` can't check anywhere in `schema`, else undefined. */
export function unchecked(schema: unknown): string | undefined {
  try {
    walk(schema);
    return undefined;
  } catch (error) {
    if (error instanceof TypeError) return error.message;
    throw error;
  }
}

function holdsIn(schema: unknown, value: unknown, root: Json): boolean {
  if (typeof schema === "boolean") return schema;
  if (!isObject(schema)) throw new TypeError(`not a schema: ${String(schema)}`);
  if ("if" in schema) {
    const branch = holdsIn(schema["if"], value, root)
      ? schema["then"]
      : schema["else"];
    if (branch !== undefined && !holdsIn(branch, value, root)) return false;
  }
  if ("additionalProperties" in schema && !additional(schema, value, root))
    return false;
  return Object.entries(schema).every(([keyword, argument]) =>
    skipped(keyword) ? true : checkOf(keyword)(argument, value, root),
  );
}

function skipped(keyword: string): boolean {
  return (
    ANNOTATIONS.has(keyword) ||
    ["if", "then", "else", "additionalProperties"].includes(keyword)
  );
}

function checkOf(keyword: string): Check {
  const check = CHECKS.get(keyword);
  if (check === undefined)
    throw new TypeError(`unsupported schema keyword '${keyword}'`);
  return check;
}

const SCHEMA_ARG: ReadonlySet<string> = new Set([
  "not",
  "if",
  "then",
  "else",
  "items",
  "additionalProperties",
]);
const LIST_ARG: ReadonlySet<string> = new Set(["allOf", "anyOf", "oneOf"]);
const MAP_ARG: ReadonlySet<string> = new Set(["properties", "$defs"]);

function walk(schema: unknown): void {
  if (typeof schema === "boolean") return;
  if (!isObject(schema)) throw new TypeError(`not a schema: ${String(schema)}`);
  for (const [keyword, argument] of Object.entries(schema)) {
    admit(keyword, argument);
    for (const sub of subschemas(keyword, argument)) walk(sub);
  }
}

function admit(keyword: string, argument: unknown): void {
  if (keyword === "pattern") compile(argument);
  else if (keyword === "$ref") refName(argument);
  else if (keyword === "format" && !FORMATS.has(String(argument)))
    throw new TypeError(`unsupported format '${String(argument)}'`);
  else if (!CHECKS.has(keyword) && !skipped(keyword))
    throw new TypeError(`unsupported schema keyword '${keyword}'`);
}

function subschemas(keyword: string, argument: unknown): readonly unknown[] {
  if (SCHEMA_ARG.has(keyword)) return [argument];
  if (LIST_ARG.has(keyword)) return schemas(argument);
  if (MAP_ARG.has(keyword)) return Object.values(map(argument));
  return [];
}

function isObject(value: unknown): value is Json {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function schemas(argument: unknown): readonly unknown[] {
  if (!Array.isArray(argument))
    throw new TypeError("expected a list of schemas");
  return argument;
}

function map(argument: unknown): Json {
  if (!isObject(argument)) throw new TypeError("expected a map of schemas");
  return argument;
}

function jsonEqual(a: unknown, b: unknown): boolean {
  if (Array.isArray(a))
    return (
      Array.isArray(b) &&
      a.length === b.length &&
      a.every((x, i) => jsonEqual(x, b[i]))
    );
  if (isObject(a)) {
    if (!isObject(b)) return false;
    const keys = Object.keys(a);
    return (
      keys.length === Object.keys(b).length &&
      keys.every((k) => k in b && jsonEqual(a[k], b[k]))
    );
  }
  return a === b;
}

/** The `$defs` entry a reference names, or undefined for `#`, the whole schema. */
function refName(ref: unknown): string | undefined {
  if (ref === "#") return undefined;
  if (typeof ref !== "string" || !ref.startsWith(DEFS))
    throw new TypeError(`unsupported $ref '${String(ref)}'`);
  return ref.slice(DEFS.length);
}

function ref(argument: unknown, value: unknown, root: Json): boolean {
  const name = refName(argument);
  if (name === undefined) return holdsIn(root, value, root);
  const defs = map(root["$defs"] ?? {});
  if (!(name in defs))
    throw new TypeError(`$ref to a missing definition '${name}'`);
  return holdsIn(defs[name], value, root);
}

function compile(argument: unknown): RegExp {
  if (typeof argument !== "string")
    throw new TypeError("a pattern is a string");
  try {
    return new RegExp(argument);
  } catch {
    throw new TypeError(`pattern '${argument}' doesn't compile`);
  }
}

function format(argument: unknown, value: unknown): boolean {
  const check = FORMATS.get(String(argument));
  if (check === undefined)
    throw new TypeError(`unsupported format '${String(argument)}'`);
  return typeof value !== "string" || check(value);
}

function number(argument: unknown): number {
  if (typeof argument !== "number") throw new TypeError("a bound is a number");
  return argument;
}

const TYPES: ReadonlyMap<string, (value: unknown) => boolean> = new Map<
  string,
  (value: unknown) => boolean
>([
  ["object", isObject],
  ["array", Array.isArray],
  ["string", (v) => typeof v === "string"],
  ["boolean", (v) => typeof v === "boolean"],
  ["null", (v) => v === null],
  ["integer", (v) => typeof v === "number" && Number.isInteger(v)],
  ["number", (v) => typeof v === "number"],
]);

function type(argument: unknown, value: unknown): boolean {
  const names: readonly unknown[] = Array.isArray(argument)
    ? argument
    : [argument];
  return names.some((name) => {
    const is = typeof name === "string" ? TYPES.get(name) : undefined;
    if (is === undefined)
      throw new TypeError(`unknown JSON type '${String(name)}'`);
    return is(value);
  });
}

type Compare = (a: number, b: number) => boolean;

function bound(compare: Compare): Check {
  return (argument, value) => {
    const limit = number(argument);
    return typeof value !== "number" || compare(value, limit);
  };
}

/** A string counts code points, as JSON Schema does; an array its items, an object its keys. */
function sizeOf(value: unknown): number | undefined {
  if (typeof value === "string") return [...value].length;
  if (Array.isArray(value)) return value.length;
  return isObject(value) ? Object.keys(value).length : undefined;
}

function size(of: "string" | "array" | "object", compare: Compare): Check {
  return (argument, value) => {
    const limit = number(argument);
    const n = TYPES.get(of)?.(value) === true ? sizeOf(value) : undefined;
    return n === undefined || compare(n, limit);
  };
}

function additional(schema: Json, value: unknown, root: Json): boolean {
  if (!isObject(value)) return true;
  const declared = map(schema["properties"] ?? {});
  return Object.keys(value)
    .filter((key) => !(key in declared))
    .every((key) => holdsIn(schema["additionalProperties"], value[key], root));
}

const ge: Compare = (a, b) => a >= b;
const le: Compare = (a, b) => a <= b;

const CHECKS: ReadonlyMap<string, Check> = new Map<string, Check>([
  ["not", (arg, value, root) => !holdsIn(arg, value, root)],
  ["allOf", (arg, v, root) => schemas(arg).every((s) => holdsIn(s, v, root))],
  ["anyOf", (arg, v, root) => schemas(arg).some((s) => holdsIn(s, v, root))],
  [
    "oneOf",
    (arg, v, root) =>
      schemas(arg).filter((s) => holdsIn(s, v, root)).length === 1,
  ],
  ["const", (arg, value) => jsonEqual(arg, value)],
  ["enum", (arg, value) => schemas(arg).some((e) => jsonEqual(e, value))],
  [
    "required",
    (arg, value) =>
      !isObject(value) || schemas(arg).every((k) => String(k) in value),
  ],
  [
    "properties",
    (arg, value, root) =>
      !isObject(value) ||
      Object.entries(map(arg)).every(
        ([key, sub]) => !(key in value) || holdsIn(sub, value[key], root),
      ),
  ],
  ["type", type],
  [
    "items",
    (arg, value, root) =>
      !Array.isArray(value) || value.every((item) => holdsIn(arg, item, root)),
  ],
  ["$ref", ref],
  [
    "pattern",
    (arg, value) => typeof value !== "string" || compile(arg).test(value),
  ],
  ["format", format],
  [
    "multipleOf",
    (arg, value) => typeof value !== "number" || multipleOf(value, number(arg)),
  ],
  ["minimum", bound(ge)],
  ["maximum", bound(le)],
  ["exclusiveMinimum", bound((a, b) => a > b)],
  ["exclusiveMaximum", bound((a, b) => a < b)],
  ["minLength", size("string", ge)],
  ["maxLength", size("string", le)],
  ["minItems", size("array", ge)],
  ["maxItems", size("array", le)],
  ["minProperties", size("object", ge)],
]);
